ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java

```diff
@@ -0,0 +1,411 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.query;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Monitor;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.pull.PullQueryRow;
+import io.confluent.ksql.execution.pull.StreamedRowTranslator;
+import io.confluent.ksql.rest.entity.ConsistencyToken;
+import io.confluent.ksql.rest.entity.StreamedRow;
+import io.confluent.ksql.util.ConsistencyOffsetVector;
+import io.confluent.ksql.util.KeyValue;
+import io.confluent.ksql.util.KeyValueMetadata;
+import io.confluent.ksql.util.KsqlHostInfo;
+import io.confluent.ksql.util.RowMetadata;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import io.vertx.core.impl.ConcurrentHashSet;
+import io.vertx.core.impl.FutureFactoryImpl;
+import io.vertx.core.spi.FutureFactory;
+import io.vertx.core.streams.WriteStream;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.List;
+import java.util.OptionalInt;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * This {@link WriteStream} allows for results to be streamed back to the client when
+ * running pull queries. Streaming behavior is important when dealing with large result
+ * sets since we don't want to hold the entire set in memory at once.
+ *
+ * <p>This class has non-blocking writes so that it can work both with callers that implement
+ * backpressure and those that don't. If the caller does not implement backpressure, they
+ * should block when {@link #writeQueueFull()} by calling {@link #awaitCapacity(long, TimeUnit)}.
+ */
+@SuppressWarnings("checkstyle:ClassDataAbstractionCoupling")
+public class PullQueryWriteStream implements WriteStream<List<StreamedRow>>, BlockingRowQueue {
+
+  private static final int DEFAULT_SOFT_CAPACITY = 50;
+  private static final FutureFactory FUTURE_FACTORY = new FutureFactoryImpl();
+
+  private final OptionalInt hardLimit;
+  private final StreamedRowTranslator translator;
+
+  // this monitor allows us to implement blocking reads on the poll()
+  // methods and also is used to ensure thread safety if multiple threads
+  // are using this WriteStream
+  private final Monitor monitor = new Monitor();
+  private final Monitor.Guard hasData = monitor.newGuard(() -> !isEmpty());
+  private final Monitor.Guard hasCapacity = monitor.newGuard(() -> !writeQueueFull());
+
+  @GuardedBy("m")
+  private final Queue<HandledRow> queue = new ArrayDeque<>();
+  @GuardedBy("m")
+  private int totalRowsQueued = 0;
+
+  @GuardedBy("m")
+  private boolean closed = false;
+  @GuardedBy("m")
+  private int softLimit = DEFAULT_SOFT_CAPACITY;
```

Comment:
I find the soft limit and hard limit names a little confusing since it implies that they are both limiting the same thing with two thresholds.  Really, the soft limit is the size of the buffer at this moment whereas hard limit is the total number of rows that have passed through.  Not a big deal, but curious what you think.

Reply:
I'll give it a thought and see if I can rename it anything clearer.